carma_wm/include/carma_wm/WMTestLibForGuidance.h

```diff
@@ -476,6 +477,64 @@ inline void setRouteByIds (std::vector<lanelet::Id> lanelet_ids, std::shared_ptr
   setRouteByLanelets(lanelets, cmw);
 }
 
+
+
+
+/**
+ * \brief Method adds a traffic light to the provided world model instance
+ *        NOTE: The stop line for the light will be located at the end of the owning_lanelet and formed from its two bound end points 
+ * 
+ * \param cmw The world model instance to update
+ * \param light_id The lanelet id to use for the generated traffic light regulatory element. This id should NOT be present in the map prior to this method call
+ * \param owning_lanelet_id The id of the lanelet which will own thr traffic light element. This id MUST be present in the map prior to this method being called
+ * \param controlled_lanelet_ids A list of lanelet ids which the traffic light will control access to. The ids MUST be present in the map. Additionally, they should be listed in order along the direction of travel.
+ * \param timeing_plan Optional parameter that is the timing plan to use for the light. The specifications match those of CarmaTrafficLightState.setStates()
+ *                     The default timing plan is 4sec yewllow, 20sec red, 20sec green
+ */ 
+inline void addTrafficLight(std::shared_ptr<carma_wm::CARMAWorldModel> cmw, lanelet::Id light_id, lanelet::Id owning_lanelet_id, std::vector<lanelet::Id> controlled_lanelet_ids, 
+std::vector<std::pair<ros::Time, lanelet::CarmaTrafficLightState>> timing_plan =
+{
+  std::make_pair<ros::Time, lanelet::CarmaTrafficLightState>(ros::Time(0), lanelet::CarmaTrafficLightState::PERMISSIVE_MOVEMENT_ALLOWED), // Just ended green
+  std::make_pair<ros::Time, lanelet::CarmaTrafficLightState>(ros::Time(4.0), lanelet::CarmaTrafficLightState::PERMISSIVE_CLEARANCE), // 4 sec yellow
+  std::make_pair<ros::Time, lanelet::CarmaTrafficLightState>(ros::Time(24.0), lanelet::CarmaTrafficLightState::STOP_AND_REMAIN), // 20 sec red
+  std::make_pair<ros::Time, lanelet::CarmaTrafficLightState>(ros::Time(44.0), lanelet::CarmaTrafficLightState::PERMISSIVE_MOVEMENT_ALLOWED) // 20 sec green
+}) {
+  
+  std::vector<lanelet::Lanelet> controlled_lanelets;
+
+  for (auto id : controlled_lanelet_ids) {
+    auto iterator = cmw->getMutableMap()->laneletLayer.find(id);
+    
+    if (iterator == cmw->getMutableMap()->laneletLayer.end())
+      throw std::invalid_argument("Provided with lanelet id not in map: " + std::to_string(id));
+
+    controlled_lanelets.push_back(*iterator);
+  }
+
+  // Get owning lanelet by id
+  auto iterator = cmw->getMutableMap()->laneletLayer.find(owning_lanelet_id);
+    
+  if (iterator == cmw->getMutableMap()->laneletLayer.end())
```

Comment:
I wasn't able to find the definition of this find, but would it work better to initialize this iterator to -1 and check against that here?

Reply:
find is defined here and uses iterator and const_iterator the same way std::vector works. 
https://github.com/usdot-fhwa-stol/autoware.ai/blob/c51ccf194c704cc2335a37d1a8235837e024fad1/lanelet2/lanelet2_core/include/lanelet2_core/LaneletMap.h#L90

I do not agree with initializing it to -1. The correct practice with iterators is to compare them with container.end() as an iterator is an object and is not equivalent to an index though the index can be extracted from it.

Reply:
Ok that sounds good then. Thanks for the explanation!